pkgs/node/src/node.zig

```diff
@@ -711,6 +864,199 @@ test "Node peer tracking on connect/disconnect" {
+    const allocator = arena_allocator.allocator();
+
+    var ctx = try testing.NodeTestContext.init(allocator, .{});
+    defer ctx.deinit();
+
+    var mock = try networks.Mock.init(allocator, ctx.loopPtr(), ctx.loggerConfig().logger(.mock), null);
+    defer mock.deinit();
+
+    const backend = mock.getNetworkInterface();
+
+    const chain_config = ctx.takeChainConfig();
+    const anchor_state = ctx.takeAnchorState();
+    var mock_chain = try stf.genMockChain(allocator, 3, ctx.genesisConfig());
+    defer mock_chain.deinit(allocator);
+    try ctx.signBlockWithValidatorKeys(allocator, &mock_chain.blocks[1]);
+    try ctx.signBlockWithValidatorKeys(allocator, &mock_chain.blocks[2]);
+
+    const test_registry = try allocator.create(NodeNameRegistry);
+    defer allocator.destroy(test_registry);
+    test_registry.* = NodeNameRegistry.init(allocator);
+    defer test_registry.deinit();
+
+    var node: BeamNode = undefined;
+    try node.init(allocator, .{
+        .config = chain_config,
+        .anchorState = anchor_state,
+        .backend = backend,
+        .clock = ctx.clockPtr(),
+        .validator_ids = null,
+        .nodeId = 0,
+        .db = ctx.dbInstance(),
+        .logger_config = ctx.loggerConfig(),
+        .node_registry = test_registry,
+    });
+    defer node.deinit();
+
+    // Create a chain of blocks: genesis -> block1 -> block2
+    // We'll cache block2 (missing block1), then when block1 arrives,
+    // processCachedDescendants should process block2. Blocks are generated
+    // via the block builder so signatures, state roots, and proposer data are valid.
+    const block1 = mock_chain.blocks[1];
+    const block2 = mock_chain.blocks[2];
+    const block1_root = mock_chain.blockRoots[1];
+    const block2_root = mock_chain.blockRoots[2];
+    const block1_slot: usize = @intCast(block1.message.block.slot);
+    const block2_slot: usize = @intCast(block2.message.block.slot);
+
+    // Cache block2 (which will fail to process because block1 is missing)
+    const block2_ptr = try allocator.create(types.SignedBlockWithAttestation);
+    try types.sszClone(allocator, types.SignedBlockWithAttestation, block2, block2_ptr);
+    try node.network.cacheFetchedBlock(block2_root, block2_ptr);
+
+    // Verify block2 is cached
+    try std.testing.expect(node.network.hasFetchedBlock(block2_root));
+
+    // Verify block2 is not in the chain yet
+    try std.testing.expect(!node.chain.forkChoice.hasBlock(block2_root));
+
+    // Advance forkchoice time to block1 slot and add block1 to the chain
+    try node.chain.forkChoice.onInterval(block1_slot * constants.INTERVALS_PER_SLOT, false);
+    const missing_roots1 = try node.chain.onBlock(block1, .{});
+    defer allocator.free(missing_roots1);
+
+    // Verify block1 is now in the chain
+    try std.testing.expect(node.chain.forkChoice.hasBlock(block1_root));
+
+    // Now call processCachedDescendants with block1_root. This should discover
```

Comment:
do any cases need more rigorous testing? or any other pending test cases, create an issue and assign to mercy